woocommerce/google-listings-and-ads
# Allow backorder product availability

### Changes proposed in this Pull Request:



As part of #738.

This PR adds a `backorder` availability for the products when back-ordering is enabled for a product and its stock status is on backorder (less than or equal to zero).

This also includes some refactoring, mainly applied to the admin inputs and attributes. A new set of input classes are created that each represents the input for an attribute. This way, the attribute's label, and description are defined in their input, which seems more logical given that they're only used in the Admin and not anywhere else the attributes are used.

A new DateTime input type is introduced that uses a set of two HTML inputs (`input type="date"` and `input type="time"`) to get the availability date and time from the user. I first tried to use a third-party DateTime picker library for this but decided to go native HTML instead because these inputs are supported by most modern browsers and by using them we avoid adding yet another library (albeit small) to our package.

This DateTime input assumes that the time entered by the user is based on the site's local timezone and not the browser's; therefore, the time is always converted to UTC when stored in the database and then back to the site's timezone when displaying to the user.

### Compatibility with [WooCommerce Pre-Orders](https://woocommerce.com/products/woocommerce-pre-orders/)

Note that since this PR was the base for the Pre-Order PR (https://github.com/woocommerce/google-listings-and-ads/pull/959) and the other one got merged, any code related to Pre-Orders is previously reviewed and can be safely ignored. Although it would be a good idea to test these two functionalities together. 

For example, since the pre-orders integration hooks into the `woocommerce_gla_product_attribute_values` action, it will override the availability date set by the user in the GLA section if pre-orders is enabled for the product and an availability date value is set in the Pre-Orders extension.

### Detailed test instructions:

1. Create a new product or edit an existing one
2. From `Inventory` enable `Manage Stock?` and `Allow Backorders?` for that product
3. Make sure it has a negative or zero stock quantity
4. Under the GLA tab of the product edit page, set the availability date to a valid date. You can also set the time but it's optional. Time is set in the site's timezone but stored in DB in UTC, so please check that the timezone conversion applies correctly.
5. Update the product and wait for it to be synced to Merchant Center
6. Confirm that the product's availability in MC is now set to "Backorder" instead of "In Stock" and the availability date and time are set correctly (also check the timezone).
7. Remove the availability date and update the product
8. Confirm that the availability date is deleted from product metadata and MC

**Test compatibility with WooCommerce Pre-Orders**

1. Install and activate [WooCommerce Pre-Orders](https://woocommerce.com/products/woocommerce-pre-orders/)
2. Enable pre-orders for the product and set an availability date via the Pre-Orders extension
3. Set a different availability date in the GLA section and also allow back orders for the product
4. Update the product and wait for it to sync
5. Confirm that the product availability is now set to "preorder" and its availability date is the one set via the Pre-Orders extension


### Changelog entry

> Add - Allow backorder stock availability for products
